### PR TITLE
Migrated to React 0.14.3

### PR DESCRIPTION
**Pull request by johnnyreilly**
This PR ports react 0.13.3 to 0.14.3 which should handle the issues raised in https://github.com/Microsoft/TypeScriptSamples/issues/61

All works fine / tests pass. One quirk; I'm having to capture the React variable in each of the JSX tests and in the `main.tsx` like so:

```
import * as React from 'react';

// ....

const __react = React; // only in place to prevent React being purged from dependencies as not used directly
```

These files do not directly use the imported `React` and so, without the workaround, React is not loaded. I think this is the TypeScript compiler performing an optimisation here; saying "well, they didn't use React directly so they must not need it".  I initially thought this was a problem with ts-loader but came to the conclusion it was probably tsc related.  Take a look at [this demo in the playground](http://www.typescriptlang.org/Playground#src=import%20'.%2Fdependencies'%3B%0Aimport%20*%20as%20React%20from%20'react'%3B%0Aimport%20*%20as%20ReactDOM%20from%20'react-dom'%3B%0Aimport%20App%20from%20'.%2Fcomponents%2FApp'%3B%0A%2F%2Fconst%20__react%20%3D%20React%3B%20%2F%2F%20only%20in%20place%20to%20prevent%20React%20being%20purged%20from%20dependencies%20as%20not%20used%20directly%0A%0AReactDOM.render(%3CApp%20%2F%3E%2C%20document.getElementById('content'))%3B%0A). You can see that React is not considered a dependency by TypeScript unless the `__react = React` is uncommented.

@jbrantly would you be able to advise if there is a better way around this than what I've done here? (I know you pretty much handled the port of React typings to 0.14 so I thought it might be worth asking.)

Just to be clear, the PR is safe to merge but it would be good to find out if the workaround could be replaced with something a little nicer.  I suspect it can't

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@mhegazy and @RyanCavanaugh

Thanks @johnnyreilly!

This isn't an optimization as much as it is us trying to give the user more control over this. If we didn't perform this step, there'd be no way to say "I only need the typings from this import." On the other hand, with the current behavior, you can always force an import by

* Adding an extra blind import:
    ```ts
    import * as React from "react";
    import "react";
    ```

* Using an imported entity as an expression (like in the following statement expression):
    ```ts
    import * as React from "react";
    React;
    ```

Actually, would you mind shortening your fix to one of the above forms just because it's a bit cleaner?

**Comment by johnnyreilly** [comment]
Done.  I had to use approach 2 as approach 1 didn't work.  Any idea why that might be?

**Comment by DanielRosenwasser** [comment]
That's weird - approach 1 didn't work?

I pulled down the changes and tried using an `import "react";` and just running `tsc` from `src.. The compiled output maintained the import. If I'm not mistaken, this sounds like a potential problem with the other tools. A sanity check from @mhegazy and @RyanCavanaugh would help.

**Comment by johnnyreilly** [comment]
Curious - ts-loader is doing the TSC invoking so @jbrantly might have a view.

**Comment by johnnyreilly** [comment]
Try approach 1 on the repo and run `npm run serve` to see the issue

File: es6-babel-react-flux-karma/test/components/Greeting.tests.tsx [hunk 1]
```diff
@@ -1,8 +1,8 @@
-import * as React from 'react/addons';
+import * as React from 'react';
+import * as TestUtils from 'react-addons-test-utils';
 import Greeting from '../../src/components/Greeting';
 import * as GreetingActions from '../../src/actions/GreetingActions';
-
-const { TestUtils } = React.addons;
+React; // use React as an expression to prevent React being purged from dependencies as not used directly
```
Comment (mhegazy) [inline es6-babel-react-flux-karma/test/components/Greeting.tests.tsx]:
```ts
import "react";
``` ?

File: es6-babel-react-flux-karma/test/components/App.tests.tsx [hunk 2]
```diff
@@ -1,10 +1,10 @@
-import * as React from 'react/addons';
+import * as React from 'react';
+import * as TestUtils from 'react-addons-test-utils';
 import App from '../../src/components/App';
 import WhoToGreet from '../../src/components/WhoToGreet';
 import Greeting from '../../src/components/Greeting';
 import GreetingStore from '../../src/stores/GreetingStore';
-
-const { TestUtils } = React.addons;
+React; // use React as an expression to prevent React being purged from dependencies as not used directly
```
Comment (mhegazy) [inline es6-babel-react-flux-karma/test/components/App.tests.tsx]: i would just add an explicit import to react:
```ts
import "react";
```

File: es6-babel-react-flux-karma/test/components/WhoToGreet.tests.tsx [hunk 3]
```diff
@@ -1,8 +1,8 @@
-import * as React from 'react/addons';
+import * as React from 'react';
+import * as TestUtils from 'react-addons-test-utils';
 import WhoToGreet from '../../src/components/WhoToGreet';
 import * as GreetingActions from '../../src/actions/GreetingActions';
-
-const { TestUtils } = React.addons;
+React; // use React as an expression to prevent React being purged from dependencies as not used directly
```
Comment (mhegazy) [inline es6-babel-react-flux-karma/test/components/WhoToGreet.tests.tsx]: same here.

File: es6-babel-react-flux-karma/src/main.tsx [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import './dependencies';
-import * as React from 'react/addons';
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
 import App from './components/App';
+React; // use React as an expression to prevent React being purged from dependencies as not used directly
```
Comment (mhegazy) [inline es6-babel-react-flux-karma/src/main.tsx]: i would just add an explicit import to react:
```ts
import "react";
```

**Comment by mhegazy** [comment]
Just one comment, using `import "react"` instead of the use as value.

File: es6-babel-react-flux-karma/test/components/Greeting.tests.tsx (hunk 1 above)
Comment (johnnyreilly) [inline es6-babel-react-flux-karma/test/components/Greeting.tests.tsx]: nope `Can't find variable: React` results

File: es6-babel-react-flux-karma/test/components/WhoToGreet.tests.tsx (hunk 3 above)
Comment (johnnyreilly) [inline es6-babel-react-flux-karma/test/components/WhoToGreet.tests.tsx]: nope `Can't find variable: React` results

File: es6-babel-react-flux-karma/test/components/App.tests.tsx (hunk 2 above)
Comment (johnnyreilly) [inline es6-babel-react-flux-karma/test/components/App.tests.tsx]: nope `Can't find variable: React` results

File: es6-babel-react-flux-karma/src/main.tsx (hunk 4 above)
Comment (johnnyreilly) [inline es6-babel-react-flux-karma/src/main.tsx]: nope `Can't find variable: React` results

**Comment by jbrantly** [comment]
@johnnyreilly I've never run into the issue I don't think because my JSX is always within a React component (so extending `React.Component`).

I wonder if the lack of `jsx` compiler option in the test tsconfig.json has anything to do with it. Per @RyanCavanaugh [here](https://github.com/Microsoft/TypeScript/issues/3764#issuecomment-119237001) when using `jsx: "react"` any JSX is supposed to act as if `React` was referenced. I can understand why that might not be the case for `jsx: "preserve"`.

**Comment by johnnyreilly** [comment]
Hi @mhegazy 

> Just one comment, using import "react" instead of the use as value.

I did try this but it didn't work I'm afraid. Please feel free to prove me wrong!

**Comment by DanielRosenwasser** [comment]
I've repro'd the issue on my side as well, it's just that somewhere in the pipeline, the `import "react"` isn't getting respected.

**Comment by johnnyreilly** [comment]
Import can't get no respect? That makes Aretha sad :smile:

**Comment by jbrantly** [comment]
The difference between `React;` and `import "react"` can also be seen in the playground example given earlier by @johnnyreilly. If you use `React;` then React is both added to the list of dependencies AND injected into the module factory function. If you use `import "react";` then it's added to the list of dependencies but not injected.

This actually makes sense to me. When I think of `import "react"` I'm thinking of importing a module for a side-effect, not somehow saying that a completely different imported variable (albeit from the same module) is somehow "used".

I don't think there is an actual bug here, although it's a bit of a weird situation. This is handled correctly with `--jsx react` because TS knows that the JSX is being compiled to `React.createElement` expressions so it knows to keep the imported variable. With `--jsx preserve` there's no guarantee that React is being used at all.

I think the correct course of action is to either use `React;` or just use `--jsx react`.

File: es6-babel-react-flux-karma/src/main.tsx (hunk 4 above)
Comment (mhegazy) [inline es6-babel-react-flux-karma/src/main.tsx]: i thought `React` is added to the global scope when you import `react`..

**Comment by mhegazy** [comment]
:+1:

**Comment by mhegazy** [comment]
thanks!

**Comment by johnnyreilly** [comment]
Pleasure. When Babel fix the sourcemaps in v6 I intend to upgrade the sample to use Babel 6 instead of v5. Just waiting on a PR to be merged here:

https://github.com/babel/babel/pull/3108

File: es6-babel-react-flux-karma/test/components/App.tests.tsx (hunk 2 above)
Comment (mhegazy) [inline es6-babel-react-flux-karma/test/components/App.tests.tsx]: @johnnyreilly with https://github.com/Microsoft/TypeScript/pull/6290, this should not be needed now. the React import should be preserved regardless. do you want to send a PR for it?

Comment (johnnyreilly) [inline es6-babel-react-flux-karma/test/components/App.tests.tsx]: Sweet - yes will do @mhegazy

**Pull request merged**